File: src/test/java/com/vorise/leetcode/n213_rob.java
package com.vorise.leetcode;

/**
 * https://leetcode.cn/problems/house-robber-ii/description/
 *
 * 213. 打家劫舍 II
 * 
 * 你是一个专业的小偷，计划偷窃沿街的房屋，每间房内都藏有一定的现金。这个地方所有的房屋都 围成一圈
 * ，这意味着第一个房屋和最后一个房屋是紧挨着的。同时，相邻的房屋装有相互连通的防盗系统，如果两间相邻的房屋在同一晚上被小偷闯入，系统会自动报警 。
 * 
 * 给定一个代表每个房屋存放金额的非负整数数组，计算你 在不触动警报装置的情况下 ，今晚能够偷窃到的最高金额。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 输入：nums = [2,3,2]
 * 输出：3
 * 解释：你不能先偷窃 1 号房屋（金额 = 2），然后偷窃 3 号房屋（金额 = 2）, 因为他们是相邻的。
 * 示例 2：
 * 
 * 输入：nums = [1,2,3,1]
 * 输出：4
 * 解释：你可以先偷窃 1 号房屋（金额 = 1），然后偷窃 3 号房屋（金额 = 3）。
 * 偷窃到的最高金额 = 1 + 3 = 4 。
 * 示例 3：
 * 
 * 输入：nums = [1,2,3]
 * 输出：3
 * 
 * 
 * 提示：
 * 
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 1000
 *
 * @author zugle on 2023-08-26
 */
public class n213_rob {

}

class Solution {

    // 动态规划，环形n-2,n-1比较
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }

        return Math.max(rob(nums, 0, nums.length - 2), rob(nums, 1, nums.length - 1));
    }

    private int rob(int[] nums, int left, int right) {
        int pre2 = 0;
        int pre1 = 0;
        for (int i = left; i <= right; i++) {
            int cur = Math.max(pre2 + nums[i], pre1);
            pre2 = pre1;
            pre1 = cur;
        }
        return pre1;
    }
}

File: src/problem/n213_rob.rs
struct Solution {}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        // 动态规划，环形n-2,n-1比较
        let size = nums.len();
        if size == 0 {
            return 0;
        }
        if size == 1 {
            return nums[0];
        }

        let s1 = Solution::rob2(&nums, 0, size - 2);
        let s2 = Solution::rob2(&nums, 1, size - 1);
        return s1.max(s2);

    }

    pub fn rob2(nums: &Vec<i32>, left: usize, right: usize) -> i32 {
        let mut pre1=0;
        let mut pre2=0;
        for (i, v) in  nums.iter().enumerate() {
            if i < left {
                continue;
            }
            if i > right {
                break;
            }

            let cur = pre1.max(pre2 + v);
            pre2 = pre1;
            pre1 = cur;
        }
        return pre1;
    }
}
